File: sdks/js/anthropic/cases/9-message-truncation.js
/**
 * 9-message-truncation: Message Truncation
 *
 * Tests that large messages are properly truncated while preserving the original message count.
 * Sends multiple large messages (~9KB each) and verifies that Sentry captures:
 * - gen_ai.request.messages.original_length (the actual count of messages sent)
 * - gen_ai.request.messages array (potentially truncated for telemetry)
 * - The relationship: len(messages) <= original_length
 */

const { Sentry } = require("../setup");
const Anthropic = require("@anthropic-ai/sdk");
const { runTestCase } = require("../../_test-utils/test-runner.cjs");

async function testLogic(inputs) {
  const { model, message_size_kb, message_count } = inputs;

  // Initialize Anthropic client
  const anthropic = new Anthropic({
    apiKey: process.env.ANTHROPIC_API_KEY,
  });

  // Generate large message content (~9KB each)
  // Using approximately 1 character per byte for ASCII
  const contentSize = message_size_kb * 1024;
  const largeContent = "A".repeat(contentSize);

  // Build messages array with alternating user/assistant messages
  // Anthropic requires conversation to alternate between user and assistant
  const messages = [];
  for (let i = 0; i < message_count; i++) {
    if (i % 2 === 0) {
      // User messages
      messages.push({
        role: "user",
        content: `Message ${i + 1}: ${largeContent}`,
      });
    } else {
      // Assistant messages (for conversation history)
      messages.push({
        role: "assistant",
        content: `Message ${i + 1}: ${largeContent}`,
      });
    }
  }

  // Ensure the last message is from user (required by Anthropic)
  if (messages[messages.length - 1].role === "assistant") {
    messages.push({
      role: "user",
      content: "Please summarize what we discussed.",
    });
  }

  // Make the API call with large messages
  const message = await anthropic.messages.create({
    model: model,
    max_tokens: 1024,
    messages: messages,
  });

  if (!message.content || message.content.length === 0) {
    throw new Error("No completion returned from Anthropic");
  }

  if (process.env.SENTRY_AI_TEST_VERBOSE === "true") {
    console.log(`    Response: ${JSON.stringify(message.content[0])}`);
    console.log(
      `    Sent ${messages.length} messages with ~${message_size_kb}KB each`
    );
  }
}

module.exports = runTestCase("9-message-truncation", testLogic, Sentry);

File: sdks/py/openai-agents/cases/9-message-truncation.py
"""
9-message-truncation: Message Truncation Test

Tests that when large messages are sent to an LLM via OpenAI Agents, Sentry
correctly tracks the original message count vs. the potentially truncated
message count in the captured span data.
"""

import os
from agents import Agent, Runner
from test_runner import run_test_case


async def test_logic(inputs):
    """The actual test logic"""
    model = inputs["model"]
    message_size_kb = inputs.get("message_size_kb", 9)
    message_count = inputs.get("message_count", 3)

    # Generate large content for the prompt (~9KB each message)
    large_content = "x" * (message_size_kb * 1024)

    # Build a large prompt with multiple "messages" embedded
    prompt_parts = []
    for i in range(message_count):
        prompt_parts.append(f"Message {i + 1}: {large_content}")

    large_prompt = (
        "\n\n".join(prompt_parts) + "\n\nPlease summarize the above messages briefly."
    )

    agent = Agent(
        name="summarizer",
        instructions="You are a helpful assistant that summarizes content.",
        model=model,
    )

    result = await Runner.run(agent, large_prompt)

    if not result.final_output:
        raise Exception("No output returned from OpenAI Agents")

    if os.getenv("SENTRY_AI_TEST_VERBOSE") == "true":
        print(f"    Response: {result.final_output[:100]}...")
        print(
            f"    Sent prompt with {message_count} large messages (~{message_size_kb}KB each)"
        )

    return result.final_output


# Export test case functions
test_case = run_test_case("9-message-truncation", test_logic)
main = test_case["main"]
assert_sentry = test_case["assert_sentry"]
